lib/charms/kubernetes_charm_libraries/v0/volumes.py

```diff
@@ -0,0 +1,415 @@
+            return
+        try:
+            statefulset = self.client.get(
+                res=StatefulSet, name=statefulset_name, namespace=self.namespace
+            )
+        except ApiError:
+            raise KubernetesRequestedVolumesError(f"Could not get statefulset {statefulset_name}")
+        containers: list[Container] = statefulset.spec.template.spec.containers  # type: ignore[attr-defined]  # noqa: E501
+        requested_volumes_mounts = [
+            requested_volume.volume_mount for requested_volume in requested_volumes
+        ]
+        container = next(container for container in containers if container.name == container_name)
+        container.volumeMounts = [
+            item for item in container.volumeMounts if item not in requested_volumes_mounts
+        ]
+        if any(
+            [
+                requested_volume.volume.emptyDir.medium == "HugePages"
+                for requested_volume in requested_volumes
+            ]
+        ):
+            try:
+                del container.resources.limits["hugepages-1Gi"]
+                del container.resources.requests["hugepages-1Gi"]
+            except KeyError:
+                pass
+        statefulset_volumes = statefulset.spec.template.spec.volumes  # type: ignore[attr-defined]  # noqa: E501
+        requested_volumes_volumes = [
+            requested_volume.volume for requested_volume in requested_volumes
+        ]
+        statefulset.spec.template.spec.volumes = [  # type: ignore[attr-defined]
+            item for item in statefulset_volumes if item not in requested_volumes_volumes
+        ]
+        try:
+            self.client.replace(  # type: ignore[call-overload]
+                name=statefulset_name,
+                obj=statefulset,
+                namespace=self.namespace,
+                field_manager=self.__class__.__name__,
+            )
+        except ApiError:
+            raise KubernetesRequestedVolumesError(
+                f"Could not replace statefulset {statefulset_name}"
+            )
+        logger.info("Requested volumes removed from %s statefulset", statefulset_name)
+
+
+class KubernetesVolumesPatchLib:
+    """Class to be instantiated by charms requiring requested volumes."""
+
+    def __init__(
+        self,
+        namespace: str,
+        application_name: str,
+        unit_name: str,
+        container_name: str,
+    ):
+        """Constructor for the KubernetesVolumesPatchLib.
+
+        Args:
+            namespace: Namespace name
+            application_name: Charm application name
+            unit_name: Unit name
+            container_name: Container name
+        """
+        self.kubernetes = KubernetesClient(namespace=namespace)
+        self.model_name = namespace
+        self.application_name = application_name
+        self.unit_name = unit_name
+        self.container_name = container_name
+
+    def add_requested_volumes(self, requested_volumes, container_name) -> None:
```

Comment:
I'm not sure about the user experience here, I don't think we want the charm to have to say to "add" or "remove" resources. The charm should simply tell the lib what it wants and the lib should figure out on its own what to do. I would recommend to have a similar approach then the Multus lib.

Just like with Multus, we will probably want to inherit from from ops.framework's Object, have the class take a "hugepages_volume_names_func" (or something along those lines) and observe config changed event

Reply:
I had this discussion with @Gmerold and decided to treat this library as a standard library and not a charm library, so avoiding inheritance from `Object` and observing events from here, but only exposing methods to add/remove volumes and handling them from the charm using the library.

Reply:
Why? Is there a good reason for multus to be different than hugepages? If this is a charm library, it can contain most of the hugepages related logic.

Reply:
In multus lib, we delete all NADs that are provisioned in K8S but are not in the list returned by the `network_attachment_definitions_func` (implemented in the UPF operator). Applying the same exact logic for volumes may have detrimental effects, as we may end up deleting all volumes mounted in the pod.
I propose an hybrid approach: the function `required_volumes_func` will return two iterables, `volumes_to_add` and `volumes_to_delete` (both defaulted to an empty list), and the logic will be implemented in the UPF operator. When HugePages are enabled, then `volumes_to_add` will contain the HugePages volumes and mounts and `volumes_to_delete` will be empty; when HugePages are disabled, then `volumes_to_add` will be empty and `volumes_to_delete` will contain the HugePages volumes to be deleted.
Let me know if you think this works fine (it may also improve the user experience on the operator side).

Reply:
If they are not charm libraries, we will end up having logic around the management of multus NAD's and hugepages volumes in both the charm and the lib. It makes the life of the charm author more difficult when they want to use the lib. I'd rather have the lib contain all of this logic and the charm only instantiate 1 object. This is what we do with `KubernetesServicePatch`, `MetricsEndpointProvider`, `KubernetesMultusCharmLib` and a bunch more, I see no reason to change this here. 

As for the patching vs creating new ones, I'm not sure honestly, this should be validated. In any case, I don't belive the outcome of this question will change my opinion around the lib being a charm library. In both cases we'd be able to provide a "volumes_with_hugepages" to the object.